torchvision/models/swin_transformer.py

```diff
@@ -0,0 +1,516 @@
+    Args:
+        dim (int): Number of input channels.
+        num_heads (int): Number of attention heads.
+        window_size (int): Window size. Default: 7.
+        shift_size (int): Shift size for shifted window attention. Default: 0.
+        mlp_ratio (float): Ratio of mlp hidden dim to embedding dim. Default: 4.0.
+        dropout (float): Dropout rate. Default: 0.0.
+        attention_dropout (float): Attention dropout rate. Default: 0.0.
+        stochastic_depth_prob: (float): Stochastic depth rate. Default: 0.0.
+        norm_layer (nn.Module): Normalization layer.  Default: nn.LayerNorm.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int,
+        window_size: int = 7,
+        shift_size: int = 0,
+        mlp_ratio: float = 4.0,
+        dropout: float = 0.0,
+        attention_dropout: float = 0.0,
+        stochastic_depth_prob: float = 0.0,
+        norm_layer: Callable[..., nn.Module] = nn.LayerNorm,
+    ):
+        super().__init__()
+
+        self.norm1 = norm_layer(dim)
+        self.attn = ShiftedWindowAttention(
+            dim,
+            window_size,
+            shift_size,
+            num_heads,
+            attention_dropout=attention_dropout,
+            dropout=dropout,
+        )
+        self.stochastic_depth = StochasticDepth(stochastic_depth_prob, "row")
+        self.norm2 = norm_layer(dim)
+        self.mlp = MLPBlock(dim, int(dim * mlp_ratio), dropout)
+
+    def forward(self, x: Tensor):
+        x = x + self.stochastic_depth(self.attn(self.norm1(x)))
+        x = x + self.stochastic_depth(self.mlp(self.norm2(x)))
+        return x
+
+
+class SwinTransformer(nn.Module):
+    """
+    Implements Swin Transformer from the `"Swin Transformer: Hierarchical Vision Transformer using
+    Shifted Windows" <https://arxiv.org/pdf/2103.14030>`_ paper.
+    Args:
+        patch_size (int): Patch size. Default: 4.
+        num_classes (int): Number of classes for classification head. Default: 1000.
+        embed_dim (int): Patch embedding dimension. Default: 96.
+        depths (List(int)): Depth of each Swin Transformer layer. Default: [2, 2, 6, 2].
+        num_heads (List(int)): Number of attention heads in different layers. Default: [3, 6, 12, 24].
+        window_size (int): Window size. Default: 7.
+        mlp_ratio (float): Ratio of mlp hidden dim to embedding dim. Default: 4.0.
+        dropout (float): Dropout rate. Default: 0.0.
+        attention_drop_rate (float): Attention dropout rate. Default: 0.0.
+        drop_path_rate (float): Stochastic depth rate. Default: 0.0.
+        block (nn.Module, optional): SwinTransformer Block. Default: None.
+        norm_layer (nn.Module, optional): Normalization layer. Default: None.
+    """
+
+    def __init__(
+        self,
+        patch_size: int = 4,
+        num_classes: int = 1000,
+        embed_dim: int = 96,
+        depths: List[int] = [2, 2, 6, 2],
+        num_heads: List[int] = [3, 6, 12, 24],
```

Comment:
We shouldn't use mutable values as default values. If `depths` or `num_heads` get modified, due to the way that Python works, their default values will change for all new instances of the class. Effectively the default values are shared for all instances of the class and if one instance modifies it, it affects all others. This is why we often opt for the `some_value = None` idiom in this codebase.

I recommend the following:
1. Most of the parameters that change across variants such as depths, num_heads and embed_dim (excluding stochastic_depth_prob) should be mandatory and thus have no default values.
2. I would shuffle things around to make them required and I would rearrange the order of the args to make them aligned with ViT and other implementations (for example num_classes is typically added prior the bock/norm_layer

Reply:
thanks a lot, I totally agree with your advice.